Publish to opensensemap

fel115 opened the pull request:
@markus2330: Which data should be published to `opensesemap.org`? And do I need to create an account for this task?

markus2330 commented:
> @markus2330: Which data should be published to opensesemap.org?

Everything of the weather station.

> And do I need to create an account for this task?

I created one for you but I did not create the box yet, please do that.

fel115 commented:
@markus2330 I tried to reproduce the function `send_message_once` from the `nextcloud.rs` file. But if I try to implement it in my code, I always get some compiling errors, because I don't use async functions. 
```
expected opaque type, found enum `Result`
```

Have you any idea what's the problem here? Have you had the same issue?

fel115 commented:
> @markus2330 I tried to reproduce the function `send_message_once` from the `nextcloud.rs` file. But if I try to implement it in my code, I always get some compiling errors, because I don't use async functions.
> 
> ```
> expected opaque type, found enum `Result`
> ```
> 
> Have you any idea what's the problem here? Have you had the same issue?

NVM. I now used the same versions of reqwest, tokio and futures for testing. And now it's compiling and working. :relaxed:

markus2330 commented:
> So it should be like a register dump? But with the extension to upload the data? How should I store the registers, in a csv file?

I am not sure if I understand. With the weather station you will always read the registers. So the difference is always only what is afterwards done with the data. I hope I could clarified the different use cases in #50. Ideally we should write down the use cases explicitly as done in PermaplanT, see https://github.com/ElektraInitiative/PermaplanT/tree/master/doc/usecases

fel115 commented:
I think usecases would be a good idea . 👍

Now I'm saving the registers, the opensensemap-id, factor and datatype of the sensors in a constant, should this be moved to a CSV file or something else? As I know Elektra is only a key:value database, so we can't store this information there. Or should we stay with the constance's?

The box-id an access token could be stored in Elektra or is this a task for later?

fel115 commented:
It should be mergeable now. But I couldn't test it on `messstation`, because of #51. CI should also run without errors.

markus2330 commented on doc/Opensensemap.md:
```diff
@@ -0,0 +1,20 @@
+### 2.1 Choose a name for your station
+### 2.2 Choose exposure (`indoor/outdoor/mobile`)
+### 2.3 Set Latitude and Longitude or select location in map
+### 2.4 Set Hardware, with `Manual configuration`
```
What hardware to add for clima_sensor_us?

fel115 replied:
It is possible to use hardware that is known from opensensemap, but to configure other sensors you need to set all sensors manually.  So config option is called Hardware.

markus2330 replied:
There are endless possibilities but we only implement and document the one hardware with all its features (until someone shows up and needs something else). Please describe here exactly what you changed in the Web-Interface, so that all sensor data gets displayed.

markus2330 commented on doc/Opensensemap.md:
```diff
@@ -0,0 +1,20 @@
+## 3. Add Sensors
+
+### 3.1 Choose a icon
+### 3.2 Set a name/phenomenon, unit and type (use of const reg-name in clima_sensor_us.rs)
```
at least what is in the brackets could be normal text and not in a title?

markus2330 commented on files/opensesame.spec:
```diff
@@ -126,3 +126,14 @@ description=How often to get new data (default: every 60 seconds, which is the h
+type = boolean
+default = 0
+
+[opensensemap/box_id]
```
Probably put below weatherstation or do you think there is something else to publish? (I don't think so.)

```suggestion
[weatherstation/opensensemap/id]
```

markus2330 commented on files/opensesame.spec:
```diff
@@ -126,3 +126,14 @@ description=How often to get new data (default: every 60 seconds, which is the h
+default = 0
+
+[opensensemap/box_id]
+description = Which opensensemap Box should be connected to the weather station
```
```suggestion
description = Which opensensemap Box should be connected to the weather station, see doc/Opensensemap.md
```

markus2330 commented on files/opensesame.spec:
```diff
@@ -126,3 +126,14 @@ description=How often to get new data (default: every 60 seconds, which is the h
+
+
+[opensensemap/access_token]
+description = Access-Token for opensensemap Box
```
```suggestion
description = Access-Token for opensensemap box
```

markus2330 commented on files/opensesame.spec:
```diff
@@ -126,3 +126,14 @@ description=How often to get new data (default: every 60 seconds, which is the h
+description = Which opensensemap Box should be connected to the weather station
+
+
+[opensensemap/access_token]
```
Avoid separators other than /:

```suggestion
[weatherstation/opensensemap/token]
```

markus2330 commented on src/main.rs:
```diff
@@ -533,6 +536,24 @@ fn main() -> Result<(), Error> {
+		if config.get_bool("weatherstation/enable") {
+			match weather_station.handle() {
+				Ok(TempWarningStateChange::ChangeToCloseWindow) => {
+					nc.send_message(gettext!("Temperature above {}°C, close the window", 23));
```
Please also translate texts.

markus2330 commented on src/weather_station/connection_modbus.rs:
```diff
@@ -1,40 +1,28 @@
-/*This program doesn't work, because there is no implementation of rtu_set_custom_rts in libmodbus-rs*/
+/*This program tests if the connection with the weather station is possible and for that reason the register 40015 is read.*/
```
This program eventually should read all registers and dump it to stdout (redirect to file can be done by the user).

fel115 replied:
Yes, I think that's a good idea.

fel115 replied:
Done in PR #52

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+/// Before using this module you need to configure elektra with the following elements
+/// [opensensemap/box_id] and [opensensemap/access_token]
```
Isn't weatherstation/enable needed, too?

fel115 replied:
I implemented the enable in the `main.rs`, should it be implemented in the `clima_sensor_us.rs`?

markus2330 replied:
I was only referring to the incomplete documentation here.

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+const REG_ROLL_MAGNETIC_COMPASS_EW: u16 = 0x8917;
+
+//OpenSenseMap
+const UPDATE_FREQU: u32 = 30000; // 30000ms*10ms = 5min
```
Is it written like this in the data sheet? If it is, you can keep it as-is, otherwise I would prefer fully-written names:

```suggestion
const UPDATE_FREQUENCY: u32 = 30000; // 30000ms*10ms = 5min
```

fel115 replied:
No, I used `UPDATE_FREQUENCY` to read all registered every 5 min. Instead of calling them all 10 ms. So it is nothing form the data sheet.

markus2330 replied:
And we probably should use 1 minute, as the web-interface from opensensemaps suggests that updates happen every minute.

markus2330 replied:
One important change needed: the weatherstation needs to be integrated in the sensors loop, so that messstation can get both weatherstation and air quality data. The frequency also needs to be adapted to make this work. (The sensors loop gets called much less frequently).

(Later we can remove all loops and use tokio instead but for now let us keep these two loops.)

fel115 replied:
Is the sensor loop, where `environment.handle()` is called?

markus2330 replied:
yes, e.g. line 236, and writeln in 240 could be modified to log some of the weather data also on the HD.

fel115 replied:
But for me, it looks like line 236 isn't called in an endless loop. I thought this is the initialization of the env sensor.

fel115 replied:
Okay, so, in that case, the while loop (main.rs line 299) is never called and the other functions are implied disabled.

markus2330 replied:
Yes, in the sensors mode everything else is currently disabled.

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+	fn init(&mut self) {
+		self.ctx = Some(
+			Modbus::new_rtu(DEVICE, BAUDRATE, PARITY, DATA_BITS, STOP_BITS)
+				.expect("Error accured while creating new RTU Object"),
```
Typo is several times:

```suggestion
				.expect("Error occurred while creating new RTU Object"),
```

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+	use super::*;
+
+	#[test]
+	#[ignore]
```
disabled? Please also add a test with JSON data.

fel115 replied:
I disabled it because for testing I need to mock the modbus connection and for testing the JSON I also need to mock the modbus connection.

markus2330 replied:
You might be able to have a function that only returns JSON without publishing it.

fel115 replied:
Yes, but the JSON function also reads the register from the weather station. Or should I read the register in an extra function and save the register values in an array?

markus2330 replied:
If the refactoring doesn't make sense by itself (shorter functions), it is also okay to not have a JSON generation test.

fel115 replied:
created two separate methods, one for sending the data to opensensemap and one for getting data from the weather station and making a JSON format

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+		match &self.ctx {
+			Some(conn) => {
+				//first create a JSON format for sending Data
+				let mut json_payload: String = "[".to_string();
```
creation of JSON data could be in its own function (publish_to_opensensemap is too long)

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+				json_payload.pop();
+				json_payload.push_str(&"]");
+
+				//Send Jason to https://api.opensensemap.org
```
```suggestion
				//Send JSON to https://api.opensensemap.org
```

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,523 @@
+					match self.publish_to_opensensemap() {
+						Ok(_) => {}
+						Err(error) => {
+							eprintln!("publishing to opensensemap failed with {}", error);
```
Modules shouldn't print, instead the error should be propagated to main and there it should be sent to nextcloud (translated string).

markus2330 commented on doc/Opensensemap.md:
```diff
@@ -0,0 +1,20 @@
+# Opensensemap
+## 1. Setup Account
+
```
Text missing?

fel115 replied:
I meant  the creation of an account, which I think, is self explaining.

markus2330 replied:
Then write something like:

```suggestion

After you've setup your account and logged in on [opensensemap.org](https://www.opensensemap.org/), you need to do following steps:

```

markus2330 reviewed (changes requested):
Great job, good to have some data already online. Looking forward to see it constantly running soon.

markus2330 commented:
@fel115 please resolve conflicts/rebase so that the CI can check cargo clippy

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,549 @@
+/// Before using this module you need to configure elektra with the following elements
```
```suggestion
/// Before using this module you need to configure Elektra with the following elements
```

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,549 @@
+			TempWarning::WarningTempNoWind
+		));
+		assert!(
+			clima_sens.set_warning_active(33.0, 3.5) == TempWarningStateChange::ChangeToCloseWindow
```
This shouldn't be triggered again. Warnings only can go one level up, but are only completely cancelled (they don't go down one by one).

markus2330 commented on lang/opensesame.pot:
```diff
@@ -61,130 +61,150 @@ msgstr ""
+msgstr ""
+
+#: src/main.rs:246
+msgid "Temperature above {}°C an no Wind"
```
```suggestion
msgid "Temperature above {}°C and no Wind"
```

markus2330 commented on lang/de_AT.po:
```diff
@@ -64,139 +64,161 @@ msgstr "Ein unbehebbarer Fehler ist bei {}:{}: {} passiert"
+msgstr "Es hat über {}°C, schließen Sie das Fenster"
+
+#: src/main.rs:246
+msgid "Temperature above {}°C an no Wind"
```
(should be regenerated, fix in source code)

```suggestion
msgid "Temperature above {} °C and no Wind"
```

markus2330 commented on lang/de_AT.po:
```diff
@@ -64,139 +64,161 @@ msgstr "Ein unbehebbarer Fehler ist bei {}:{}: {} passiert"
+msgstr "Es hat über {}°C und es geht kein Wind"
+
+#: src/main.rs:249
+msgid "Temperature above {}°C"
```
```suggestion
msgid "Temperature above {} °C"
```

markus2330 commented on lang/de_AT.po:
```diff
@@ -64,139 +64,161 @@ msgstr "Ein unbehebbarer Fehler ist bei {}:{}: {} passiert"
+
+#: src/main.rs:249
+msgid "Temperature above {}°C"
+msgstr "Es hat über {}°C"
```
```suggestion
msgstr "Es hat über {} °C"
```

markus2330 commented on lang/de_AT.po:
```diff
@@ -64,139 +64,161 @@ msgstr "Ein unbehebbarer Fehler ist bei {}:{}: {} passiert"
+msgstr "Es hat über {}°C"
+
+#: src/main.rs:252
+msgid "Temperature again under {}°C, remove warning"
```
```suggestion
msgid "Temperature again under {} °C, warning was removed"
```

markus2330 commented on lang/de_AT.po:
```diff
@@ -64,139 +64,161 @@ msgstr "Ein unbehebbarer Fehler ist bei {}:{}: {} passiert"
+
+#: src/main.rs:252
+msgid "Temperature again under {}°C, remove warning"
+msgstr "Es hat unter {}°C, die Warnung wird entfernt"
```
```suggestion
msgstr "Es hat unter {} °C, die Warnung wird entfernt"
```

markus2330 reviewed (changes requested):
It is getting better :rocket: I found many little things and a logic error.

fel115 commented on doc/Opensensemap.md:
```diff
@@ -0,0 +1,63 @@
+## 3. Obtaining IDs
+
+### 3.1 senseBox-ID
+Find the senseBox-ID in the `Dashboard` section.
```
No, I used mine, but it is possible to transfer the box to another account. I can generate a token that you insert in your account and then the box belongs to your account.

markus2330 replied:
Yes, please send me the token via email.

fel115 commented on doc/Opensensemap.md:
````diff
@@ -0,0 +1,63 @@
+kdb set user:/sw/libelektra/opensesame/#0/current/weatherstation/opensensemap/token "<opensensemap-access-token>"
+```
+### 4.3 Enabling Sensor Mode
+Enable sensor mode by creating a sensor profile with the following settings:
````
@markus2330 should I add here the workaround with `kdb editor ...`?

markus2330 replied:
If the code here does not work, write something that does work or simply use pseudo-code and leave it up to the user (I'll use Ansible anyway).

fel115 commented on src/main.rs:
```diff
@@ -234,6 +237,40 @@ fn main() -> Result<(), Error> {
+							"Error durring handling weather station: {}",
+							error.to_string()
+						));
+						writeln!(
+							&mut outfile,
+							"{} weather station: {}",
+							Local::now().format(&date_time_format).to_string(),
+							error.to_string(),
+						)
+						.unwrap();
```
The Env also outputs its values in a new line. Should we really use `write!`?

fel115 replied:
In this line I only output, if an error occurs. The sensor data need to be printed in the `create_json` of the clima_sensor_us module.

markus2330 replied:
If it is only error, then don't write it all into this file. The file should only contain valid data.

markus2330 replied:
If it is only error, then don't write it all into this file. The file should only contain valid measurement data.

In the long term, I want to get rid of this file anyway.

markus2330 commented on src/main.rs:
```diff
@@ -234,6 +237,40 @@ fn main() -> Result<(), Error> {
+			if enable_weather_station {
+				match weather_station.handle() {
+					Ok(TempWarningStateChange::ChangeToCloseWindow) => {
+						nc.send_message(gettext!("Temperature above {}°C, close the window", 23));
```
Simply a search in the Internet for emojis or unicode (together with the term of what you want).

markus2330 commented on doc/Opensensemap.md:
```diff
@@ -18,17 +18,28 @@ This option is utilized for manually configuring the weather station's sensors.
+
+A constant within the [source code](../src/clima_sensor_us.rs) takes the form: `const REG_AIR_TEMP: u16 = 0x88BD;`
+
+In this illustrative scenario, the variable `REG_AIR_TEMP` represents the air temperature. Our initial action involves transferring the parameter designation `Air temperature` to the `Phenomenon` input section. Subsequently, we proceed to incorporate `°C` from the provided table into the `Unit` input section. Finally, the variable name `REG_AIR_TEMP` is copied into the `Type` input field.
```
```suggestion
In this illustrative scenario, the variable `REG_AIR_TEMP` represents the air temperature.
Our initial action involves transferring the parameter designation `Air temperature` to the `Phenomenon` input section.
Subsequently, we proceed to incorporate `°C` from the provided table into the `Unit` input section.
Finally, the variable name `REG_AIR_TEMP` is copied into the `Type` input field.
```

markus2330 commented on src/main.rs:
```diff
@@ -240,33 +240,29 @@ fn main() -> Result<(), Error> {
-						nc.send_message(gettext!("Temperature above {}°C, close the window", 23));
+						nc.send_message(gettext!(
+							"🌡️ Temperature above {} °C, close the window",
+							23
```
Use constants for these numbers.

markus2330 reviewed (changes requested):
Logic error does not seem to be fixed yet? Please send token and please always rerequest review once you are finished. The cycles should become faster to get enough done in August.

markus2330 commented:
Please also pump version in Cargo.toml/lock here, so that we know if correct version is used. On messstation now an old version might be running again?)

fel115 commented:
> Logic error does not seem to be fixed yet? Please send token and please always rerequest review once you are finished. The cycles should become faster to get enough done in August.

The logic error should be removed already. I added a new test line to check it, but it seems to work. I implemented it so that the states can move up, but only down if the `RemoveWarning` temperature is reached.

Which version do you mean? The Cargo.toml/lock in this PR are up to date with my local version.

horenso commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,560 @@
+		s
+	}
+
+	fn init(&mut self) {
```
init should return a `Result<(), SomeErrorType>` so that we don't have to `expect()` everything here. If you want to panic if init fails we can do that in the caller.

horenso commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,560 @@
+			opensensemap_counter: 0,
+		};
+		if config.get_bool("weatherstation/enable") {
+			s.init();
```
So panic here if `init()` returns error, add a TODO and make init() return a `Result` would be the easiest for now,
of of course make new() return a `Result` as well.

fel115 replied:
What do you mean with 'add a TODO'?

horenso replied:
Sorry if that was unclear. I would suggest we make both `new` and `init` return a `Result`. Then in `main` you properly want to panic for now if `new()` returns an error. So in `main` we just do this:

```
let mut weather_station = ClimaSensorUS::new(&mut config).expect("Could not create weather station!");
```

Then we only use `expect` in one place ^^

horenso commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,560 @@
+		{
+			new_warning = TempWarning::WarningTempNoWind;
+		} else if temp > 23.0
+			&& !matches!(self.warning_active, TempWarning::WarningTemp)
+			&& !matches!(self.warning_active, TempWarning::WarningTempNoWind)
```
```suggestion
			&& !matches!(self.warning_active, TempWarning::WarningTemp | TempWarning::WarningTempNoWind)
```

you can match on multiple variants here

fel115 replied:
Thanks for that information, looks much cleaner. :smiley:

horenso commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,560 @@
+			&& !matches!(self.warning_active, TempWarning::WarningTempNoWind)
+		{
+			new_warning = TempWarning::CloseWindow;
+		} else if !matches!(self.warning_active, TempWarning::None)
+			&& !matches!(self.warning_active, TempWarning::RemoveWarning)
```
```suggestion
		} else if !matches!(self.warning_active, TempWarning::None | TempWarning::RemoveWarning)
```

horenso reviewed (changes requested):
Looks good, I just suggested some minor things to fix or add TODOs.

markus2330 reviewed (changes requested):
Version not pumped as requested, didn't look further. Please also fix all other problems.

markus2330 commented:
CI also fails, please always check this before requesting reviews.

fel115 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,560 @@
+							}
+							if value != ERROR_CODE_S32 as f32 && value != ERROR_CODE_U32 as f32 {
+								json_payload.push_str(&format!(
+									"{}\"sensor\":\"{}\",\"value\":\"{}\"{},",
```
Thanks for that info, I'll create a TODO

fel115 commented:
> Version not pumped as requested, didn't look further. Please also fix all other problems.

What is missing?

> CI also fails, please always check this before requesting reviews.

Yes, I requested a review yesterday and did changes today. But now it should work.

markus2330 commented:
> What is missing?

Pumping (i.e. increasing) the version number of opensesame (in [Cargo.toml](https://github.com/ElektraInitiative/opensesame/pull/49/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542) and Cargo.lock) Please always ask if something is unclear.

> Yes, I requested a review yesterday and did changes today.

Let us better do it the other way round next time.

For me the deb from the build server on messstation does not work (I also tried restarting):

```
Failed to init libmodbus connection: Rtu { msg: "rtu_set_serial_mode", source: Os { code: 95, kind: Uncategorized, message: "Operation not supported" } }
```

Maybe something in `debian/service` is causing the problem (NoNewPrivileges=true or ProtectSystem=full)? Any ideas?

Btw. there are spelling errors in the PR. Let us use some spell checker to make this easier for both of us: #64

fel115 commented:
> Pumping (i.e. increasing) the version number of opensesame (in [Cargo.toml](https://github.com/ElektraInitiative/opensesame/pull/49/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542) and Cargo.lock) Please always ask if something is unclear.

Now I understand it :sweat_smile: 

> For me the deb from the build server on messstation does not work (I also tried restarting):
> 
> ```
> Failed to init libmodbus connection: Rtu { msg: "rtu_set_serial_mode", source: Os { code: 95, kind: Uncategorized, message: "Operation not supported" } }
> ```
> 
> Maybe something in `debian/service` is causing the problem (NoNewPrivileges=true or ProtectSystem=full)? Any ideas?


I encountered this error only when I switched from using `RtuRS232` to `RtuRS485` mode. It's possible that an additional privilege is required, which might not be configured during the Opensesame installation. The system functions perfectly without Opensesame installed. The specific permission we require is for `rtu_set_custom_rts(RequestToSendMode::RtuRtsUp)` in order to manipulate the GPIO pins.

markus2330 commented:
There is a conflict, please fix.

The requirements changed a bit, see #65 (22 instead of 23 etc.).

>   I encountered this error only when I switched from using RtuRS232 to RtuRS485 mode. It's possible that an additional privilege is required, which might not be configured during the Opensesame installation. The system functions perfectly without Opensesame installed. The specific permission we require is for rtu_set_custom_rts(RequestToSendMode::RtuRtsUp) in order to manipulate the GPIO pins.

Could you please find out on messstation which permission it is?

I now claimed the device, hopefully it continues working. Sensebox ID is 64cb602193c69500072a580f

Will review later, I am not sure if I can today. But without the permission error fixed, the weather station is not finished anyway. Would love to see it running as normal installed service :rocket:

fel115 commented:
> There is a conflict, please fix.

Is done.

> The requirements changed a bit, see #65 (22 instead of 23 etc.).

I will take a look at it, after I fixed the issue with the `.deb` installation.

> I now claimed the device, hopefully it continues working. Sensebox ID is 64cb602193c69500072a580f

I tested it by starting opensesame in debug mode and it is still working.

> Could you please find out on messstation which permission it is?
> Will review later, I am not sure if I can today. But without the permission error fixed, the weather station is not finished anyway. Would love to see it running as normal installed service 🚀

I tried to change the permission but it didn't change anything. It's really strange, I have no idea why it isn't working. I compiled a release of the opensesame version and with that, it works perfectly, but by installing the service nothing is working. For now, I am stuck, I have no idea what I could do. I changed the service-file to run as root and it didn't work. I tried to execute it manually and it didn't work. Could it be something while creating the `.deb` file ?

markus2330 commented on src/clima_sensor_us.rs:
```diff
@@ -0,0 +1,569 @@
+			&& !matches!(self.warning_active, TempWarning::WarningTemp)
+		{
+			new_warning = TempWarning::WarningTempNoWind;
+		} else if temp > 23.0
```
Constants not used here, and should be >= 22.

markus2330 commented on src/main.rs:
```diff
@@ -205,9 +212,11 @@ fn main() -> Result<(), Error> {
 	let mut remember_baseline_counter = 0;
 	let wait_for_remember_baseline = 300000 * 24 * 7; // 7 days
+	let enable_weather_station = config.get_bool("weatherstation/enable");
+	let mut weather_station =
+		ClimaSensorUS::new(&mut config).expect("Failed to init libmodbus connection");
```
Does not send error message to NC (only `Failed to init libmodbus connection`)

markus2330 commented on src/main.rs:
```diff
@@ -234,6 +243,39 @@ fn main() -> Result<(), Error> {
 				handle_environment(&mut environment, &mut nc, None, &mut config);
 			}
 
+			if enable_weather_station {
```
This can be easily avoided by letting handle do nothing if the weather station is not enabled.

fel115 replied:
I did it implicitly, by checking the value of the Modbus connection. If it is set to `None` no state change is returned. But should it be explicitly been checked in the `handle()` function?

markus2330 replied:
Yes, all the other modules handle this internally in handle, I don't see a reason why this module should be different in this regard. Do you already have any plans how the new module system should be? (E.g. with async handle entry point?)

fel115 replied:
Ok. 
According async I need to talk with @horenso, but the sync modules only need to be enabled or disabled in elektra config.

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -39,4 +39,4 @@ To solve this issue we had to use the right address, which is configured on the
 ### Solve
-To solve this issue we had to use the right function (`0x01`, `0x02`, ...) for the chosen register. The function for any register is in the documantation of the [weatherstation](https://www.vetterag.ch/images/pdf/thies/BA/4.920x.x0.xxx_ClimaSensor_US_d.pdf) or [inverter](https://loxwiki.atlassian.net/wiki/spaces/LOX/pages/1605274474/Goodwe+GW10+ET+MODBUS+TCP+IP?preview=/1605274474/1605274552/Good ...
\ No newline at end of file
+To solve this issue we had to use the right function (`0x01`, `0x02`, ...) for the chosen register. The function for any register is in the documentation of the [weatherstation](https://www.vetterag.ch/images/pdf/thies/BA/4.920x.x0.xxx_ClimaSensor_US_d.pdf) or [inverter](https://loxwiki.atlassian.net/wiki/spaces/LOX/pages/1605274474/Goodwe+GW10+ET+MODBUS+TCP+IP?preview=/1605274474/1605274552/Good ...
```
```suggestion
To solve this issue we had to use the right function (`0x01`, `0x02`, ...) for the chosen register.
The function for any register is in the documentation of the [weatherstation](https://www.vetterag.ch/images/pdf/thies/BA/4.920x.x0.xxx_ClimaSensor_US_d.pdf) or [inverter](https://loxwiki.atlassian.net/wiki/spaces/LOX/pages/1605274474/Goodwe+GW10+ET+MODBUS+TCP+IP?preview=/1605274474/1605274552/Goodwe_Modbus_Protocol_Hybrid_ET_EH_BH_BT__ARM205%20HV__V1.7%20_%20Read%20Only_20200226%20(1).pdf).
```

markus2330 reviewed (changes requested):
> I will take a look at it, after I fixed the issue with the .deb installation.

It is not a good approach to make unrelated bug fixes dependent on each other. On the contrary: you get better throughput of work when you fix everything that you can fix and give that way "unfixable things" a bit of a rest before you tackle them systematically.

It is possible, that #68 is not related to the code here at all. So please get this finally done.

> Could it be something while creating the .deb file ?

Could be, seems like systematic investigation is needed, please check #68.

markus2330 commented:
CI system is failing now. Let get us done before you start refactoring :rocket:

fel115 commented:
I think the error says, that the `asm/param.h` header couldn't be found.
```bash
[2023-08-21T08:16:15.708Z]   /usr/include/linux/param.h:5:10: fatal error: 'asm/param.h' file not found

[2023-08-21T08:16:15.708Z]   /usr/include/linux/param.h:5:10: fatal error: 'asm/param.h' file not found, err: true

[2023-08-21T08:16:15.708Z]   thread 'main' panicked at 'could not reate binding: ()', /usr/local/cargo/git/checkouts/libmodbus-rs-6e4ee3ccb3d4329a/ad36c57/libmodbus-sys/build.rs:126:10

[2023-08-21T08:16:15.708Z]   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

[2023-08-21T08:16:15.708Z] warning: build failed, waiting for other jobs to finish...

[2023-08-21T08:16:23.568Z] cargo-deb: build failed
```

But I found the following files on in the docker image:
```bash
/usr/include/x86_64-linux-gnu/asm/param.h
/usr/include/asm-generic/param.h
/usr/arm-linux-gnueabihf/include/asm-generic/param.h
/usr/arm-linux-gnueabihf/include/asm/param.h
```

Maybe we are using the wrong `param.h` or we need to tell cargo to use one of them. On the olimex board I found this header-file:
```bash
/usr/include/arm-linux-gnueabihf/asm/param.h
```

# 1 Attempt using `ln`
```bash
ln -s /usr/arm-linux-gnueabihf/include/asm /usr/include/arm-linux-gnueabihf/asm
```
compiling libmodbus-sys seems to work now, but I get another error.
## Following error
```bash
 Compiling libmodbus-sys v1.0.0 (https://github.com/ElektraInitiative/libmodbus-rs.git#ad36c57f)
   Compiling libmodbus v1.0.0 (https://github.com/ElektraInitiative/libmodbus-rs.git#ad36c57f)
   Compiling opensesame v0.7.9 (/home/smod/Bachelorarbeit/GitHub/opensesame)
error: linking with `arm-linux-gnueabihf-gcc` failed: exit status: 1
  |
  = note: "arm-linux-gnueabihf-gcc" "/tmp/rustcXsXwkr/symbols.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.0.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.1.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.10.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.11.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.12.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.13.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.14.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.15.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.2.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.3.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.4.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.5.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.6.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.7.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.8.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.opensesame.1f5581e4-cgu.9.rcgu.o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c.5g4y87cs9c17k125.rcgu.o" "-Wl,--as-needed" "-L" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps" "-L" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/release/deps" "-L" "/usr/lib" "-L" "/usr/lib/x86_64-linux-gnu" "-L" "/usr/lib/arm-linux-gnueabihf" "-L" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/build/libssh2-sys-132c8fccb0a99275/out/build" "-L" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/build/libz-sys-29c1330973dae8d9/out/lib" "-L" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/build/libz-sys-29c1330973dae8d9/out/lib" "-L" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib" "-Wl,-Bstatic" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsignal_hook-232c3e321de410c5.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsystemstat-b4e2528bd1581b04.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnom-eed8580659cec4ba.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbytesize-44298f8891546820.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsunrise-045d42ac772bc3eb.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libchrono-5bdd4b7f5b53a1f8.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnum_integer-d9af4c21563f09e1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnum_traits-54972ef245fb6823.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libgettextrs-75fa23ebca410aa1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libgettext_sys-01062f5b401ca7ec.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblocale_config-cd92e52d176646ed.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libregex-fa65e9025a48d00a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libaho_corasick-644f3fd55dda16e2.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libregex_syntax-952e1a9af526ba55.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libssh2-ec5de5b962ce88d3.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblibssh2_sys-ffbc572a3196bff5.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblibz_sys-5321df6e86a3890b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbme280-706788cffa5e3da7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblinux_embedded_hal-3e6f9c7cf960282b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsysfs_gpio-1fae2f38f1dfee15.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libspidev-01df6da8cbb96568.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libserial_unix-1f567f8a27eeb01e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libioctl_rs-98d6abdccaccc7a4.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtermios-202ccd0b526e6a66.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libserial_core-0873c1dc9ae3c78e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libembedded_hal-b1253051ec6130b3.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libvoid-6c237b419ece004d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnb-8a4390a674e787cd.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnb-b9f9b5241178ab27.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcast-98f93fda6d26a7f1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libreqwest-d97857a4623ef077.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libuuid-af5188890150f921.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand-1e9fc6e7d1ce885a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_xorshift-dc9ce4620cb72832.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_pcg-d0c2ef9de5fc6942.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_hc-d9081d11a184c5d4.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_chacha-fde6514b26d5896b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_isaac-358b9a660ab1d84a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_core-02c05f2c6bd2a443.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_os-4337b53b1b427e21.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_jitter-a754e8191a79c96c.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librand_core-0db65a24d2beba92.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libserde_urlencoded-08ee19a35889a939.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libdtoa-25b654185b462ddb.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmime_guess-fa6c34a3cf864bf2.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libunicase-10a7927167de8039.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmime-a157ac0f76521dd2.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libflate2-67d15530116ccc7f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcrc32fast-9298b7da189d6393.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhyper_tls-90fde798e46d8e8d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnative_tls-a1c8caee06b1253d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libopenssl_probe-e2f1f3de5776f266.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libopenssl-7c215acd94d7a7ae.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libforeign_types-6f846c8610852c05.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libforeign_types_shared-b0af12525e43f70e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libopenssl_sys-c5d9e66826d2d143.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhyper-e154a408dc0fbe2a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libwant-e35913938b96a5cd.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtry_lock-9ea20396364f9e57.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio-57eca94078682dc1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_timer-302430d54173db82.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_threadpool-7a8e4c0413a1b06d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcrossbeam_queue-3335b1cb5ee2bf03.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcrossbeam_deque-89802f250f5efb51.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcrossbeam_epoch-3138d3a1acd7d759.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmemoffset-1afb9ec3f8b51b76.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_tcp-02724e137ed93da5.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_reactor-963a1bdcfdb44e47.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_sync-286ca763348e5ae6.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libparking_lot-92d7541abcd50eb2.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libparking_lot_core-e3de6df0b0b707f7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsmallvec-9f48ec711bc55b0d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmaybe_uninit-fad483eb87242071.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblock_api-0d89eb1a405545f3.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_current_thread-cdcf4bb98f4bfc55.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_executor-60c3b3c43204acb1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcrossbeam_utils-6d49174b61ad6770.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblazy_static-dda1dbd8d119ec1b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmio-feb5a0477bed7a22.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnet2-d98bc8ac7300fb01.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhttparse-d7ec6a661285225c.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhttp_body-41e29aeb30b8ec59.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_buf-7714c5f87fefe8c0.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libh2-62814ba45bef5127.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libindexmap-c9e39a9074be48ab.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhashbrown-9424b44c6adee49f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libstring-8be36f86b2a181ac.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio_io-1871973ac07463ea.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_cpupool-20ba959fcde76030.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libhttp-3b113f82fde07c50.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libitoa-f1e2c1dc3b7b4ce3.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfnv-84e83f80b3b3522e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures-9b353229636caa56.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libencoding_rs-b32ba0e4d3ba7324.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcookie_store-3826e789d79efa77.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libserde_json-dfe4fac9d6ef71db.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libryu-d9ea470c7eb953af.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libitoa-c229e3b411822bc4.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libpublicsuffix-5058d9241e7236d7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liburl-8be891949a61b658.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libidna-3752024277933b16.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libform_urlencoded-3c46b1bcaff6e70f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libpercent_encoding-7d17486aeb87e029.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtry_from-b75dcfe2e0e8fd82.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcfg_if-5ada8c169eadbbf1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libserde-41a7c2a953dcfcb7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfailure-7c476bc947e6da74.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbacktrace-ad1d7c0fb1eb624d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libminiz_oxide-27bf8e8f6684ef0b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libadler-f776cc6a2afb43f7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libobject-fc0ec465e7ff6da5.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libaddr2line-7677990fd5459f9e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libgimli-289b3466cda4fa19.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/librustc_demangle-fce2d799dbe5bcf7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcookie-f050ddaad544cd7e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtime-88d1a738ebbaf24d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liburl-dcc64e2cc4c09d4a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libpercent_encoding-9e6b82e58c5435fb.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libidna-a6b0694d4ba0b58f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libunicode_normalization-bb61b3b15db2cb84.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtinyvec-42e3f6c8b826f9e4.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtinyvec_macros-2766c3f5188fb347.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libunicode_bidi-5d8f74cc396e407e.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmatches-6468a1faef4f9f62.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbytes-17af39cc9b40408b.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libeither-64e2b4671ec3b17d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libiovec-127cc41d9d197b43.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbase64-d2da5db1298534cb.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libi2cdev-15edf61d38f625d0.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbyteorder-805c1b090e80c28a.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libelektra-125689d337c71318.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libelektra_sys-608e754d6d703f43.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblibmodbus-9f734d91f4413eb8.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libgpio_cdev-f0823992576af32d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libtokio-48a1809d5d178084.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsignal_hook_registry-bbdadc3efaee1b67.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnum_cpus-260f2475be66116f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libonce_cell-f86e669a62206a39.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbytes-e34f0bb6dda91833.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmio-ba7198582ac4e6b1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblog-bd0a30d5e028ee76.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libparking_lot-9aefec37420e0c16.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libparking_lot_core-5d5bfb91a463def1.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libsmallvec-dedbadbdd8ad1159.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblock_api-d74c44f25ba8a4fc.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libscopeguard-1b9b0a2fccfe2513.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libinstant-c7d1111939502a80.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures-4d169f55dbf95e69.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_executor-e72a35c46f5d25d2.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_util-d0507078d01da79f.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmemchr-bae805a069f57e4c.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_io-99059cc092b025c6.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libslab-4cac2e6fff293ce7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_channel-c817d53d3b449d47.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libpin_project_lite-1c412671a1e4a0b7.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_sink-6be7bd64b4d3675c.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_task-1dbee2f00380ca37.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libpin_utils-d32ff0e468e0f0a5.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libfutures_core-5f62043f0312ed8d.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libnix-271e9131526f7e11.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libmemoffset-14d4fad7060e6abd.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libcfg_if-adc66b3e0b92b3c4.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/libbitflags-eeca23656a8dc436.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblibmodbus_sys-518ffeb1e2334027.rlib" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/liblibc-598b306b9fbe3dbf.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libstd-5441313738423f3c.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libpanic_unwind-18bb98e073b57c66.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libobject-18493e66d801dfe5.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libmemchr-55118b25e640f526.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libaddr2line-fd9aaf36d1c2adb8.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libgimli-5d58e865994d4aa1.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/librustc_demangle-aa59bec315822c4e.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libstd_detect-35a74a95e36b4dd8.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libhashbrown-1a56bc629f92cee8.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libminiz_oxide-baf247ad07828283.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libadler-73b352ed08a46f40.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/librustc_std_workspace_alloc-21c200c2a02e9373.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libunwind-5fc0c3d88f2d3564.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libcfg_if-cdffa1f684e2313b.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/liblibc-bd73747efa500215.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/liballoc-2ad560159065d348.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/librustc_std_workspace_core-b3ec66ab0c76e5be.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libcore-15bd8b817cdb1714.rlib" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib/libcompiler_builtins-56008f1ecc8f8a2f.rlib" "-Wl,-Bdynamic" "-lc" "-lc" "-lssl" "-lcrypto" "-lelektra-core" "-lelektra-meta" "-lelektra-kdb" "-lmodbus" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/usr/local/rustup/toolchains/1.67.0-x86_64-unknown-linux-gnu/lib/rustlib/armv7-unknown-linux-gnueabihf/lib" "-o" "/home/smod/Bachelorarbeit/GitHub/opensesame/target/armv7-unknown-linux-gnueabihf/release/deps/opensesame-64244b3ff678d10c" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs"
  = note: /usr/lib/gcc-cross/arm-linux-gnueabihf/10/../../../../arm-linux-gnueabihf/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libc.so when searching for -lc
          /usr/lib/gcc-cross/arm-linux-gnueabihf/10/../../../../arm-linux-gnueabihf/bin/ld: /usr/lib/x86_64-linux-gnu/libc.a: error adding symbols: file format not recognized
          collect2: error: ld returned 1 exit status
```

I am still trying, so if you @markus2330, @4ydan, or @horenso have any tips, I would be thankful.

4ydan commented on scripts/jenkins/Dockerfile_cross:
```diff
@@ -47,7 +47,8 @@ RUN apt install -y --no-install-recommends \
     libc6-dev-armhf-cross \
     gcc-arm-linux-gnueabihf \
-    binutils-arm-linux-gnueabihf
+    binutils-arm-linux-gnueabihf \
```
This is redundant I think.

fel115 replied:
So I can remove this line?

4ydan replied:
Yes, its probably already installed in some other package, but can also leave it for more robustness, idk?

markus2330 commented:
It seems to build now. Does the produced `.deb` already work? Please say when I should review.

> I think the error says

Thx, very nice analysis.

> skipping incompatible /usr/lib/x86_64-linux-gnu/libc.so

Looks like arm libc is missing?

fel115 commented:
I added the `libmodbus-dev:armhf` and now it's working. @4ydan pls review the Dockerfile.
